server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -616,8 +560,8 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         });
     }
 
-    private static ImmutableOpenMap<ShardId, ShardSnapshotStatus> processWaitingShards(
-            ImmutableOpenMap<ShardId, ShardSnapshotStatus> snapshotShards, RoutingTable routingTable) {
+    private static ImmutableOpenMap<ShardId, ShardSnapshotStatus> processWaitingShardsAndRemovedNodes(
```

Comment:
Combining both node removal and shard starting into one method saves us iterating all the shards twice for the relevant updates to each snapshot.